### PR TITLE
python312Packages.pyopencl: try to enable tests

### PR DESCRIPTION
**Pull request by GaetanLepage**
## Description of changes

- **python312Packages.pyopencl: enable tests**
- **python312Packages.pyopencl: add submodules to pythonImportsCheck**

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by GaetanLepage** [comment]
```
Executing pythonImportsCheckPhase
Check whether the following modules can be imported: pyopencl pyopencl.array pyopencl.compyte
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<string>", line 1, in <lambda>
  File "/nix/store/04gg5w1s662l329a8kh9xcwyp0k64v5a-python3-3.12.4/lib/python3.12/importlib/__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1324, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'pyopencl.array'
```

File: pkgs/development/python-modules/pyopencl/default.nix [hunk 1]
```diff
@@ -71,10 +71,11 @@ buildPythonPackage rec {
     rm -rf pyopencl
   '';
 
-  # gcc: error: pygpu_language_opencl.cpp: No such file or directory
-  doCheck = false;
-
-  pythonImportsCheck = [ "pyopencl" ];
+  pythonImportsCheck = [
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/development/python-modules/pyopencl/default.nix:74]: That file doesn't exist 😅 Do we maybe need to install it?

```
 ➜ tree result/
 result
├──  lib
│   └──  python3.12
│       └──  site-packages
│           ├──  pyopencl
│           │   └──  _cl.cpython-312-x86_64-linux-gnu.so
│           └──  pyopencl-2024.2.7.dist-info
│               ├──  licenses
│               │   └──  LICENSE
│               ├──  METADATA
│               ├──  RECORD
│               └──  WHEEL
└──  nix-support
    └──  propagated-build-inputs
```

[the remark this answers is not in the document]
Comment (GaetanLepage) [reply pkgs/development/python-modules/pyopencl/default.nix:74]: Yes maybe.
Here is what I have for a normal installation of this package with `pip`:

```
galepage@auriga:~/.local$ ls -l lib/python3.10/site-packages/pyopencl
total 1276
-rw-r--r-- 1 galepage robotlearn  51265 Aug 28 15:59 algorithm.py
-rw-r--r-- 1 galepage robotlearn 111717 Aug 28 15:59 array.py
-rw-r--r-- 1 galepage robotlearn   7960 Aug 28 15:59 bitonic_sort.py
-rw-r--r-- 1 galepage robotlearn  16172 Aug 28 15:59 bitonic_sort_templates.py
-rw-r--r-- 1 galepage robotlearn  15959 Aug 28 15:59 cache.py
-rw-r--r-- 1 galepage robotlearn   5661 Aug 28 15:59 capture_call.py
drwxr-xr-x 3 galepage robotlearn   4096 Aug 28 15:59 characterize
drwxr-xr-x 3 galepage robotlearn   4096 Aug 28 15:59 cl
-rwxr-xr-x 1 galepage robotlearn 734945 Aug 28 15:59 _cl.cpython-310-x86_64-linux-gnu.so
-rw-r--r-- 1 galepage robotlearn   8222 Aug 28 15:59 clmath.py
-rw-r--r-- 1 galepage robotlearn  13043 Aug 28 15:59 clrandom.py
-rw-r--r-- 1 galepage robotlearn   4818 Aug 28 15:59 cltypes.py
-rw-r--r-- 1 galepage robotlearn   2074 Aug 28 15:59 _cluda.py
drwxr-xr-x 4 galepage robotlearn   4096 Aug 28 15:59 compyte
-rw-r--r-- 1 galepage robotlearn  38602 Aug 28 15:59 elementwise.py
-rw-r--r-- 1 galepage robotlearn  81108 Aug 28 15:59 __init__.py
-rw-r--r-- 1 galepage robotlearn  14023 Aug 28 15:59 invoker.py
-rw-r--r-- 1 galepage robotlearn   1914 Aug 28 15:59 ipython_ext.py
-rw-r--r-- 1 galepage robotlearn    610 Aug 28 15:59 _mymako.py
drwxr-xr-x 2 galepage robotlearn   4096 Aug 28 15:59 __pycache__
-rw-r--r-- 1 galepage robotlearn  25524 Aug 28 15:59 reduction.py
-rw-r--r-- 1 galepage robotlearn  65529 Aug 28 15:59 scan.py
-rw-r--r-- 1 galepage robotlearn  45993 Aug 28 15:59 tools.py
-rw-r--r-- 1 galepage robotlearn    267 Aug 28 15:59 version.py
```

[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/development/python-modules/pyopencl/default.nix:74]: so we are missing a lot actually 😅

[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/development/python-modules/pyopencl/default.nix:74]: Is it the preBuild which deletes the files?

That depends probably on setuptools and the metadata in pyproject.toml and which directories contain a __init__.py file.

[the remark this answers is not in the document]
Comment (GaetanLepage) [reply pkgs/development/python-modules/pyopencl/default.nix]: Well, it has been fixed in https://github.com/NixOS/nixpkgs/pull/336885.
This PR has became quite minor.

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 337923` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>16 packages built:</summary>
  <ul>
    <li>python311Packages.loopy</li>
    <li>python311Packages.loopy.dist</li>
    <li>python311Packages.pyopencl</li>
    <li>python311Packages.pyopencl.dist</li>
    <li>python311Packages.reikna</li>
    <li>python311Packages.reikna.dist</li>
    <li>python311Packages.sasmodels</li>
    <li>python311Packages.sasmodels.dist</li>
    <li>python312Packages.loopy</li>
    <li>python312Packages.loopy.dist</li>
    <li>python312Packages.pyopencl</li>
    <li>python312Packages.pyopencl.dist</li>
    <li>python312Packages.reikna</li>
    <li>python312Packages.reikna.dist</li>
    <li>python312Packages.sasmodels</li>
    <li>python312Packages.sasmodels.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 337923` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>16 packages built:</summary>
  <ul>
    <li>python311Packages.loopy</li>
    <li>python311Packages.loopy.dist</li>
    <li>python311Packages.pyopencl</li>
    <li>python311Packages.pyopencl.dist</li>
    <li>python311Packages.reikna</li>
    <li>python311Packages.reikna.dist</li>
    <li>python311Packages.sasmodels</li>
    <li>python311Packages.sasmodels.dist</li>
    <li>python312Packages.loopy</li>
    <li>python312Packages.loopy.dist</li>
    <li>python312Packages.pyopencl</li>
    <li>python312Packages.pyopencl.dist</li>
    <li>python312Packages.reikna</li>
    <li>python312Packages.reikna.dist</li>
    <li>python312Packages.sasmodels</li>
    <li>python312Packages.sasmodels.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 337923` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 337923` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages marked as broken and skipped:</summary>
  <ul>
    <li>python312Packages.loopy</li>
    <li>python312Packages.loopy.dist</li>
    <li>python312Packages.pyopencl</li>
    <li>python312Packages.pyopencl.dist</li>
    <li>python312Packages.reikna</li>
    <li>python312Packages.reikna.dist</li>
    <li>python312Packages.sasmodels</li>
    <li>python312Packages.sasmodels.dist</li>
  </ul>
</details>
<details>
  <summary>8 packages failed to build:</summary>
  <ul>
    <li>python311Packages.loopy</li>
    <li>python311Packages.loopy.dist</li>
    <li>python311Packages.pyopencl</li>
    <li>python311Packages.pyopencl.dist</li>
    <li>python311Packages.reikna</li>
    <li>python311Packages.reikna.dist</li>
    <li>python311Packages.sasmodels</li>
    <li>python311Packages.sasmodels.dist</li>
  </ul>
</details>

**Review by doronbehar — APPROVED** [review]

**Pull request merged**